### ISSUE TITLE
missing string casting/folding for strands

### ISSUE THREAD

**Issue by nigoroll** [varnishcache/varnish-cache#2745]
https://github.com/varnishcache/varnish-cache/commit/d9f31889b3ff4b9a88c96f83e67be1df3f33071e exposed this regression with https://code.uplex.de/uplex-varnish/libvmod-blobdigest in this test https://code.uplex.de/uplex-varnish/libvmod-blobdigest/blob/master/src/tests/esi.vtc
```
12:21:22 ***  v1    0.5 CLI RX  106
12:21:22 **** v1    0.5 CLI RX|Message from VCC-compiler:
12:21:22 **** v1    0.5 CLI RX|Expression has type INT, expected STRANDS
12:21:22 **** v1    0.5 CLI RX|('<vcl.inline>' Line 15 Pos 49) -- (Pos 61)
12:21:22 **** v1    0.5 CLI RX|                dtop.update(blob.decode(encoded=req.esi_level));
12:21:22 **** v1    0.5 CLI RX|------------------------------------------------#############---
12:21:22 **** v1    0.5 CLI RX|
12:21:22 **** v1    0.5 CLI RX|('<vcl.inline>' Line 15 Pos 17) -- (Pos 61)
12:21:22 **** v1    0.5 CLI RX|                dtop.update(blob.decode(encoded=req.esi_level));
12:21:22 **** v1    0.5 CLI RX|----------------#############################################---
12:21:22 **** v1    0.5 CLI RX|
12:21:22 **** v1    0.5 CLI RX|Running VCC-compiler failed, exited with 2
12:21:22 **** v1    0.5 CLI RX|VCL compilation failed
12:21:22 ---- v1    0.5 FAIL VCL does not compile
12:21:22 *    top   0.5 RESETTING after ../src/tests/esi.vtc
```

**Comment by slimhazard** [comment]
Ugh. (You're decoding ESI level into a blob, srsly?)

Then we'll either have to revert that commit, or have VCC recognize STRANDS as a context in which it does string conversion. VCC already does that for STRING_LIST, I haven't looked at how, but presumably the same thing would have to happen for STRANDS.

We don't *have* to have the refactoring of VMOD blob to use STRANDS, at least not right away. We could revert it for now, if it's too much trouble before Sep 15th, and come back to it later.

I would like the VMOD to have this change, eventually anyway. It gets us to exercise STRANDS in the core code (I suppose this is the sort of thing we find out by doing that). It reduces the lines of code a bit and is IMO a cleaner foundation for the sort of thing that the VMOD does.

**Comment by nigoroll** [comment]
> Ugh.
the purpose of the test is to update a `PRIV_TOP` from various esi levels. The value being used here does not matter for the purpose of the test.

Yes, the purpose of the ticket was to note that VCC should behave identically for STRING, STRING_LIST and STRANDS with regards to casting/folding.

Regarding the options with respect to the upcoming release I think we should have a quick chat during bugwash.

Other than that, the whole point of addition STRANDS was to supersede STRING_LIST and I also want to see that transition happen. So I much appreciate your change going forward.

**Comment by slimhazard** [comment]
I've tried some changes in VCC, but am confused by the fact that VCC not only has the types STRING, STRING_LIST and STRANDS, but also STRINGS (with an 'S').

When I've tried changing VCC to include STRANDS in the string conversions, tests fail that bring about VCL compile errors, where the error message "Expression has type STRING" is expected, but it comes out has "Expression has type STRINGS".

From this I conclude that I shouldn't try to fix VCC while only partly understanding what I'm doing. And that we're better off leaving this to Monday's bugwash.
